src/imitation/algorithms/preference_comparisons.py

```diff
@@ -452,9 +746,22 @@ def __init__(
                 Default (and currently the only option) is to use synthetic preferences
                 based on ground-truth rewards. Human preferences could be implemented
                 here in the future.
+            preferences_path: path to a pickled PreferenceDataset which will be
```

Comment:
This seems like a useful feature for debugging, but I don't like how it has so many effects throughout the code, being used as a flag to toggle whether to train the agent or gather data as well.

I'm inclined to remove this, and maybe keep the changes in some private branch you could use later.

An alternative I'd be happier with is to let the callee pass in a `dataset` directly (that could be loaded from disk), have a separate flag to disable adding to the dataset (or you could even just run that code anyway but have the fixed dataset ignore any additions -- though that's inefficient; or perhaps the dataset itself could have a frozen attribute that's checked?), and a separate flag to disable agent training (or you could maybe even just make an AgentTrainer wrapper with a `train` that just forwards to `sample`).

Reply:
Agreed that it makes things unnecessarily complex, I've removed it for now.

My motivation for leaving it in was mainly that I think it will be useful if we ever have an implementation of human feedback; in that case, you might want to quickly train a reward model on already collected preferences (though this is of course not equivalent to online training, so it would just be for quick tests).

But if we decide we do need it, I think the simplest option would just be to have a script that doesn't rely on `PreferenceComparisons` at all. If we have a fixed dataset, the code to train on that is literally just
```
reward_trainer = CrossEntropyRewardTrainer(reward_net, **kwargs)
dataset = PreferenceDataset.load(preference_path)
reward_trainer.train(dataset)
```
apart from logging, creating the `RewardNet` etc. So it would be very quick to create an extra script for that if needed.
(I wanted to integrate this into the existing training script at first, but even that isn't very clean)

Reply:
(I did leave the preference saving feature in because that's not much overhead and if we ever have human feedback, I think it'll be good practice to save them)

Reply:
Ah, good point about training from an offline dataset. That is a reasonable use case, although I think one might often want to warm-start from an offline dataset but then continue to do online training. Seems tricky to support all these possible use cases, so I think breaking it down into using the underlying components is probably more scalable as you suggest.

Being able to save datasets seems good.